index.bs

```diff
@@ -1133,6 +1135,16 @@ The {{MLOperand}} objects are created by the methods of {{MLGraphBuilder}}, inte
 To <dfn for="MLGraphBuilder">validate operand</dfn> given {{MLGraphBuilder}} |builder| and {{MLOperand}} |operand|, return true if |operand|.{{MLOperand/[[builder]]}} is |builder|, and false otherwise.
 </p>
 
+#### {{MLNumber}} #### {#api-mlnumber-typedef}
+
+<dfn typedef>MLNumber</dfn> is used when specifying the type of a numeric option for an {{MLOperand}} which can be of any {{MLOperandDataType}}, including both 64-bit integer types ({{MLOperandDataType/"uint64"}} and {{MLOperandDataType/"int64"}}) and 32-bit floating point ({{MLOperandDataType/"float32"}}). Implementations must process the value according to the corresponding {{MLOperandDataType}}.
```

Comment:
> Implementations must process the value according to the corresponding {{MLOperandDataType}}.

What if there is no corresponding `MLOperandDataType`? For example, `clamp()` can be built as an `MLActivation` without needing to specify any particular dtype. What is the dtype or the activation's associated [operator](https://webmachinelearning.github.io/webnn/#dom-mlactivation-operator-slot)?

The concept of an activation's operator is a bit hand-wavy in the spec, but [it's very concrete in the Chromium implementation](https://crsrc.org/c/third_party/blink/renderer/modules/ml/webnn/ml_activation.h;drc=763ac39bd0d3d8ae0562492b4628d5ed98fc2cbf;l=40) and the data type must be known when we pass the activation as an input to some other builder method anyways (we need to check that the types match, assuming we don't want to allow mixed precision https://github.com/webmachinelearning/webnn/issues/283#issuecomment-2046275085)

Reply:
Presumably the data type of the eventual operator's input? i.e. for the impl we'd need to hold onto the union in the MLActivation until the operator is constructed?

We need to improve the spec text, but I want to understand the implementation first.

Reply:
Added a short note in 67b5a68bf88e8202f0e915d80d8c7c2def6ee504 but we can probably improve it.

Reply:
> > i.e. for the impl we'd need to hold onto the union in the MLActivation until the operator is constructed?
> 
> I think so.

Does this make sense? What happens if an `MLActivation` is passed to multiple operators with different data types?

Reply:
> What happens if an MLActivation is passed to multiple operators with different data types?

My assumption is that the MLActivation will internally need to hold the union for its whole lifetime, and the value is converted (as a copy) to a value of appropriate data type during either:
* operator creation (e.g. when conv2d() is called)
* during build, when the graph is converted to a compiled graph

Ideally the difference between those would not be observable, which might constrain how we do validation - if we throw on values (e.g. a negative number passed to a unsigned type, Infinity passed to an integer type, etc) then it would be observable if we validate during op creation or during build.

Reply:
For the sake of completeness, there's a third place where validation could happen:

- activation creation

So the levels of validation for a `clamp()` activation might look like:

- On activation creation:
  - [check that `minValue <= maxValue`](https://webmachinelearning.github.io/webnn/#check-clamp-options)
- On operator creation (with the respective activation):
  - [check `MLGraphBuilder` is the same](https://webmachinelearning.github.io/webnn/#mlgraphbuilder-validate-activation)
  - As of #646 we should have enough information to validate data types, too. This is related to the discussion of what the behavior should be if the `MLNumber` does not fit into the respective type (e.g. passing a negative number to a "uint32" operator)
- During build:
  - Currently, this includes platform-specific checks such as whether the operator supports the given activation. Moving this validation forward falls under the umbrella of #463

... this seems fine? What I find unintuitive is squaring this statement:

> the MLActivation will internally need to hold the union for its whole lifetime, and the value is converted (as a copy) to a value of appropriate data type

with the idea that an `MLActivation` has one [operator slot](https://webmachinelearning.github.io/webnn/#dom-mlactivation-operator-slot). I would assume a single "operator" would imply one data type, but then again the whole "operator" concept is pretty hand-wavy anyways so maybe this is fine :)

Reply:
To be explicit, here's where I'm envisioning the conversion/validation happens:

- for `constant()` - immediate conversion based on explicitly passed `MLOperandDataType`
- for `MLOperand clamp()` - immediate conversion based on _input_'s data type
- for `MLActivation clamp()` - when the activation passed to an operand-vending method - see 
- for `pad()` - immediate conversion based on _input_'s data type

See https://github.com/webmachinelearning/webnn/pull/649#discussion_r1575166753 for a framework where activation validation could be plugged in.

Reply:
question:
Does this mean for example for `gemm`, if the input `a` and `b` are passed as float16, and if you pass 
`{alpha: MLNumber(3.14123452435)}`, it will automatically convert that to float16 as `3.141`?

Reply:
I guess under the hood - yes (although it'd be just `{alpha: 3.14123452435}`)

Presumably that's what the spec does today, even if we didn't introduce `MLNumber`?

Reply:
Under the hood it preserves the fp32 precision right now (as what you are suggesting). 
But I'd prefer we explicitly set the expectation that it will be casted to the precision to be the same as the input type.  
In the `gemm` example, since it's emulated as `alpha * matmul(a,b) + beta * c` in CoreML, and CoreML requires all these binary ops to have the same type. so if `a`, `b` are fp16, then it's better to cast `alpha` to fp16 to multiply them. Otherwise we need to cast input to fp32 (which doesn't get executed on ANE).

So right now we just don't support fp16 inputs for gemm on CoreML until this is sorted out.